### ISSUE TITLE
rbenv section behavior

### ISSUE THREAD

**Issue by epipheus** [romkatv/powerlevel10k#333]
Am I wrong here in what I am expecting? I expected the rbenv section to only pop up if we are in a rbenv local scope of some kind. I understand there is an option to determine if you always want it to show, even when global and local are the same, but it's weird. If I have that option set to true (show even if global and local are the same, it's always on -- now that might be a weird thing to say, what did I expect? I expected it to be on when in a local scope, like a Rails app repository, like when git section pops up not like when I'm in /etc/nginx or /var/blah/blah/blah.  the option makes it always on no matter what, even in a completely different dev scope, like I dunno, my Flutter directory.  Is this expected?

**Comment by romkatv** [comment]
`rbenv` segment is shown if and only if `$POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW == true || "$(rbenv version-name)" != "$(rbenv global)"`.

Since the output of `rbenv version-name` depends on the current directory while the output of `rbenv global` does not, the visibility of `rbenv` depends on the current directory if and only if `$POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW == true`.

**Comment by epipheus** [comment]
To me still feels to always show with potentially zero ruby context present. Like if there isn't a local ruby present it should probably fall off completely.  I'd like to have it there if there is a detectable local ruby context period whether different from global or not.  But not always on no matter what, especially ina different dev context.

**Comment by romkatv** [comment]
I'm having trouble understanding your comments, both the original and the second. Could you rephrase them with more structure?

I understand that in some situation `rbenv` doesn't behave the way you want. What situation is that? How does it behave currently in this situation? How do you want it to behave in this situation?

**Comment by epipheus** [comment]
> `rbenv` segment is shown if and only if `$POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW == true || "$(rbenv version-name)" != "$(rbenv global)"`.
> 
> Since the output of `rbenv version-name` depends on the current directory while the output of `rbenv global` does not, the visibility of `rbenv` depends on the current directory if and only if `$POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW == true`.

Instead of the above, which shows the rbenv segment if always show is true or  if local ruby is different from the global version, I'm proposing that always on isn't really ideal because it really isn't contextual.  If we treated everything like that and showed everything that's global it'd get messy pretty quickly. I mean keep the option, but instead I would love to see the following.  If there is no local ruby of any kind (either only a global ruby or nothing detectable), then show nothing. However, if there is a local ruby, show it whether coincident with global or not.

How might we do that?  if `rbenv local`  has a any meaningful response, show it, any response that doesn't match "no local" then show it.  But if for some reason `rbenv local` matches `"rbenv: no local version configured for this directory"` don't show it.

That make sense?  I feel the local context if more important even for the case you care most about (where local doesn't equal global). The presence of local show probably be the driver of showing something.  Always on requiring only a global something is weird to me, I mean it would be like throwing anything in my path on the bar, right?

**Comment by romkatv** [comment]
> if `rbenv local` has a any meaningful response, show it

Thanks, this is pretty clear. I can implement this but first I would like to confirm that this is indeed what you want.

Right now `rbenv` shows either nothing or your current ruby version. For example, if prompt shows `2.6.3`, you can be certain that typing `ruby` will invoke ruby 2.6.3 and typing `ruby -v` will print `2.6.3`. `POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW` affects *when* `rbenv` prompt is shown but not *what* is shown.

If `rbenv` prompt were to always display the output of `rbenv local`, then it would sometimes show ruby version that isn't the same as what `ruby -v` prints. Here's an example:

```zsh
❯ mkdir /tmp/proj
❯ cd /tmp/proj
❯ rbenv local 2.5.3
❯ rbenv shell 2.6.3
❯ ruby -v
ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-linux]
❯ rbenv which ruby
/home/romka/.rbenv/versions/2.6.3/bin/ruby
❯ rbenv local
2.5.3
```

What would you like to happen in this situation?

1. Prompt shows `2.6.3` (the current ruby version).
2. Prompt shows `2.5.3` (the output of `rbenv local`).
3. Prompt shows nothing.

**Comment by epipheus** [comment]
I think option 1. Current Ruby. Although wow, that's kind of a crazy edge
case. I don't use shell much tho. But I think option one should be fine.
Only thing better that that would be to someone let the user define that.
Absent of that I would say option 1.

**Comment by romkatv** [comment]
I added `POWERLEVEL9K_RBENV_SOURCES` to compliment `POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW`. Here are the default values in configs produced by `p10k configure`:

```zsh
# Hide ruby version if it doesn't come from one of these sources.
typeset -g POWERLEVEL9K_RBENV_SOURCES=(shell local global)

# If set to false, hide ruby version if it's the same as global:
# $(rbenv version-name) == $(rbenv global).
typeset -g POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW=false
```

And these are the values that will give you the behavior you want:

```zsh
# Hide ruby version if it doesn't come from `rbenv shell` or `rbenv local`.
typeset -g POWERLEVEL9K_RBENV_SOURCES=(shell local)

# Don't hide ruby version even if it's the same as `rbenv global`.
typeset -g POWERLEVEL9K_RBENV_PROMPT_ALWAYS_SHOW=true
```

**Comment by epipheus** [comment]
Dude, AWESOME! Will give a try.  Thank you very much.

**Comment by epipheus** [comment]
Works beautifully.  Thanks again!

**Comment by romkatv** [comment]
Thanks for verifying 🤗
